sm_ipt/sm_ipt.c

```diff
@@ -0,0 +1,453 @@
+
+	while (true) {
+		block = queue_recv(ctx);
+		if (block < 0) {
+			break;
+		}
+
+		release = (block & FLAG_RELEASE);
+		block &= ~FLAG_RELEASE;
+
+		if (block >= NUM_BLOCKS) {
+			continue;
+		}
+
+		if (release) {
+			msg = (struct sm_ipt_message *)(&ctx->out.shm_queue->allocable[
+								block * ctx->out.block_size]);
+			blocks = (msg->size + ctx->out.block_size - 1) / ctx->out.block_size;
+			free_mask_set(ctx, calc_mask(blocks, block));
+			sm_ipt_os_give(&ctx->out_sem);
+		} else {
+			msg = (struct sm_ipt_message *)(&ctx->in.shm_queue->allocable[
+								block * ctx->in.block_size]);
+
+			if ((msg->size >= (&ctx->in.shm_queue->allocable[ctx->in.allocable_size] -
+				(uint8_t *)msg)) || (msg->size < sizeof(struct sm_ipt_message))) {
+				memory_corrupted_error();
+				break;
+			}
+
+			SM_IPT_DBG("Message received at: 0x%x, length: %d", (uint32_t)msg->buf,
+						(msg->size - sizeof(struct sm_ipt_message)));
+
+			ctx->receive_handler(msg->buf, msg->size - sizeof(struct sm_ipt_message));
+		}
+	}
+}
+
+void sm_ipt_free_rx_buf(struct sm_ipt_ctx *ctx, const uint8_t *buf)
+{
+	struct sm_ipt_message *msg = SM_IPT_OS_GET_CONTAINTER(buf, struct sm_ipt_message, buf);
+	uint32_t offset = (uint8_t *)msg - (uint8_t *)ctx->in.shm_queue->allocable;
+	uint32_t block_index = offset / ctx->in.block_size;
+
+	SM_IPT_ASSERT(block_index < NUM_BLOCKS);
+
+	queue_send(ctx, (block_index | FLAG_RELEASE));
+	SM_IPT_DBG("Dealocated RX block at 0x%x", (uint32_t)msg->buf);
+}
+
+static void handshake_step(struct sm_ipt_ctx *ctx, uint8_t this_value, uint8_t next_value)
+{
+	SM_IPT_DBG("Setting out handshake to: %d", this_value);
+
+	ctx->out.shm_queue->handshake = this_value;
+
+	SM_IPT_OS_MEMORY_BARRIER();
+
+	SM_IPT_DBG("Waiting for handshake next value: %d", next_value);
+
+	while ((ctx->in.shm_queue->handshake != this_value) &&
+		(ctx->in.shm_queue->handshake != next_value)) {
+		sm_ipt_os_yield();
+		SM_IPT_OS_MEMORY_BARRIER();
+	}
+}
+
+int sm_ipt_init(struct sm_ipt_ctx *ctx, sm_ipt_receive_handler_t callback)
+{
+	int err;
+
+	ctx->receive_handler = callback;
+
+	err = sm_ipt_os_init(&ctx->os_ctx);
+	if (err < 0) {
+		return err;
+	}
+
+	free_mask_init(ctx);
+	sm_ipt_os_sem_init(&ctx->out_sem);
+	sm_ipt_os_mutex_init(&ctx->out_mutex);
+
+	ctx->out.shm_queue = (struct sm_ipt_shm_queue *) ctx->os_ctx.out_shmem_ptr;
+	ctx->in.shm_queue = (struct sm_ipt_shm_queue *) ctx->os_ctx.in_shmem_ptr;
+
+	SM_IPT_ASSERT((uint32_t)ctx->out.shm_queue.allocable % WORD_SIZE == 0);
+	SM_IPT_ASSERT((uint32_t)ctx->in.shm_queue.allocable % WORD_SIZE == 0);
+
+	ctx->out.allocable_size = ((ctx->os_ctx.out_total_size - sizeof(struct sm_ipt_queue)) /
+							ALLOCABLE_MULTIPLY) * ALLOCABLE_MULTIPLY;
+	ctx->in.allocable_size = ((ctx->os_ctx.in_total_size - sizeof(struct sm_ipt_queue)) /
+							ALLOCABLE_MULTIPLY) * ALLOCABLE_MULTIPLY;
+	ctx->in.block_size = ctx->in.allocable_size / NUM_BLOCKS;
+	ctx->out.block_size = ctx->out.allocable_size / NUM_BLOCKS;
```

Comment:
Maybe this will make calculations more clear:

```c
ctx->in.block_size = (ctx->os_ctx.in_total_size - sizeof(struct sm_ipt_queue)) / NUM_BLOCKS / sizeof(uint32_t) * sizeof(uint32_t);
ctx->out.block_size = ...
ctx->in.allocable_size = ctx->in.block_size * NUM_BLOCKS;
ctx->out.allocable_size = ...
```

And `ALLOCABLE_MULTIPLY` is not needed any more.

Reply:
This solution is more readable, but I think WORD_SIZE should be used instead of sizeof(uint32_t).